### PR TITLE
Add unit tests for candidate_register candidate_update 1946

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1946

### CODE REVIEW

File: action/candidateregister_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+func TestCandidateRegister(t *testing.T) {
+	require := require.New(t)
+	cr, err := NewCandidateRegister(nonce, canAddress, canAddress, canAddress, canAddress, amount.Text(10), duration, autoStake, payload, gaslimit, gasprice)
+	require.NoError(err)
+
+	ser := cr.Serialize()
+	require.Equal("0a81010a29696f3178707136326177383575717a72636367397935686e727976386c64326e6b7079636333677a611229696f3178707136326177383575717a72636367397935686e727976386c64326e6b7079636333677a611a29696f3178707136326177383575717a72636367397935686e727976386c64326e6b7079636333677a611202313018e80720012a29696f3178707136326177383575717a72636367397935686e727976386c64326e6b7079636333677a6132077061796c6f6164", hex.EncodeToString(ser))
+
+	require.NoError(err)
```
Comment (CoderZhi) [inline action/candidateregister_test.go]: ?

File: action/candidateregister_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	require.Equal(gaslimit, cr.GasLimit())
+	require.Equal(gasprice, cr.GasPrice())
+	require.Equal(nonce, cr.Nonce())
+
+	require.Equal(canAddress, cr.Name())
+	require.Equal(canAddress, cr.OperatorAddress().String())
+	require.Equal(canAddress, cr.RewardAddress().String())
+	require.Equal(canAddress, cr.OwnerAddress().String())
```
Comment (CoderZhi) [inline action/candidateregister_test.go]: can we use different values?

File: action/candidateregister_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"encoding/hex"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCandidateRegister(t *testing.T) {
+	require := require.New(t)
+	cr, err := NewCandidateRegister(nonce, canAddress, canAddress, canAddress, canAddress, amount.Text(10), duration, autoStake, payload, gaslimit, gasprice)
```
Comment (CoderZhi) [inline action/candidateregister_test.go]: why do we use some values defined outside of this unit test?

File: action/candidateregister_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	require.Equal(amount, cr2.Amount())
+	require.Equal(duration, cr2.Duration())
+	require.True(cr2.AutoStake())
+	require.Equal(payload, cr2.Payload())
+}
+
+func TestCandidateRegisterSignVerify(t *testing.T) {
+	require := require.New(t)
+	require.Equal("cfa6ef757dee2e50351620dca002d32b9c090cfda55fb81f37f1d26b273743f1", senderKey.HexString())
+	cr, err := NewCandidateRegister(nonce, canAddress, canAddress, canAddress, canAddress, amount.Text(10), duration, autoStake, payload, gaslimit, gasprice)
```
Comment (CoderZhi) [inline action/candidateregister_test.go]: same problem here

File: action/candidateupdate_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	"encoding/hex"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCandidateUpdate(t *testing.T) {
+	require := require.New(t)
+	cr, err := NewCandidateUpdate(nonce, canAddress, canAddress, canAddress, gaslimit, gasprice)
```
Comment (CoderZhi) [inline action/candidateupdate_test.go]: same here

File: action/candidateregister_test.go (hunk 2 above)
Comment (lzxm160) [reply action/candidateregister_test.go]: sure

File: action/candidateregister_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	require.NoError(err)
+	require.Equal(uint64(10700), gas)
+	cost, err := cr.Cost()
+	require.NoError(err)
+	require.Equal("107010", cost.Text(10))
+
+	proto := cr.Proto()
+	cr2 := &CandidateRegister{}
+	require.NoError(cr2.LoadProto(proto))
+	require.Equal(canAddress, cr2.Name())
```
Comment (dustinxie) [inline action/candidateregister_test.go]: can simplify to Equal(cr, cr2)

File: action/candidateupdate_test.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	require.NoError(err)
+	require.Equal(uint64(10000), gas)
+	cost, err := cr.Cost()
+	require.NoError(err)
+	require.Equal("100000", cost.Text(10))
+
+	proto := cr.Proto()
+	cr2 := &CandidateUpdate{}
+	require.NoError(cr2.LoadProto(proto))
+	require.Equal(canAddress, cr2.Name())
```
Comment (dustinxie) [inline action/candidateupdate_test.go]: simplify to Equal(cr, cr2)

File: action/candidateregister_test.go (hunk 6 above)
Comment (lzxm160) [reply action/candidateregister_test.go]: not equal because of nonce gas and gas price

**Pull request merged**